### PR TITLE
Support python 3.11 changes.

### PR DESCRIPTION
**Pull request by cbalint13**
This PR adds support for required changes having python >= 3.11

---
* [Builds](https://copr.fedorainfracloud.org/coprs/rezso/ML/build/4584964/) on {py3.10, py3.11} x {aarch64, x86_64}.

Guided from here: https://docs.python.org/3.11/whatsnew/3.11.html#pyframeobject-3-11-hiding

Thanks,
~Cristian.

### CODE REVIEW

**Comment by soulitzer** [comment]
Deferring to @albanD for these changes

File: torch/csrc/autograd/profiler_python.cpp [hunk 1]
```diff
@@ -38,11 +38,18 @@ static constexpr size_t CallTypeSize = 3;
 struct CodeLocation {
   CodeLocation() = default;
   explicit CodeLocation(const PyFrameObject* frame)
+  // clang-format off
+#if PY_VERSION_HEX >= 0x030B0000
```
Comment (albanD) [inline torch/csrc/autograd/profiler_python.cpp:42]: Could we have a python compat header file that defines the macro as suggested in https://docs.python.org/3.11/whatsnew/3.11.html?
In particular, for this, we can use as suggested
```c
#if PY_VERSION_HEX < 0x030900B1
static inline PyCodeObject* PyFrame_GetCode(PyFrameObject *frame)
{
    Py_INCREF(frame->f_code);
    return frame->f_code;
}
#endif
```

That way, there is no need to have ifdef/clang-format disable everywhere in the codebase.

File: torch/csrc/Stream.cpp [hunk 2]
```diff
@@ -107,7 +107,9 @@ PyTypeObject THPStreamType = {
 
 void THPStream_init(PyObject* module) {
   THPStreamClass = &THPStreamType;
+#if PY_VERSION_HEX < 0x030B0000
```
Comment (albanD) [inline torch/csrc/Stream.cpp:110]: We should use
```c
#if PY_VERSION_HEX < 0x030900A4 && !defined(Py_SET_TYPE)
static inline void _Py_SET_TYPE(PyObject *ob, PyTypeObject *type)
{ ob->ob_type = type; }
#define Py_SET_TYPE(ob, type) _Py_SET_TYPE((PyObject*)(ob), type)
#endif
```

File: torch/csrc/autograd/python_variable.cpp [hunk 3]
```diff
@@ -573,6 +573,35 @@ static int THPVariable_clear(THPVariable* self) {
   return 0;
 }
 
+#if PY_VERSION_HEX >= 0x030B0000
+// Traverse and clear are required for supporting Python's GC cycle handling.
+static int THPFunction_traverse(THPFunction* self, visitproc visit, void* arg) {
```
Comment (albanD) [inline torch/csrc/autograd/python_variable.cpp:578]: We expect this function to be overriden and never called. I think we want this to be empty and do
```c
static int THPFunction_traverse(THPFunction* self, visitproc visit, void* arg) {
  TORCH_INTERNAL_ASSERT(false, "Tensor tp_traverse function was not overriden properly");
  return 0;
}
```

File: torch/csrc/autograd/python_variable.cpp [hunk 4]
```diff
@@ -1644,8 +1673,12 @@ PyTypeObject THPVariableType = {
     Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE |
         Py_TPFLAGS_HAVE_GC, /* tp_flags */
     nullptr, /* tp_doc */
-    // Also set by metaclass
+// Also set by metaclass
+#if PY_VERSION_HEX >= 0x030B0000
```
Comment (albanD) [inline torch/csrc/autograd/python_variable.cpp:1677]: We can set this unconditionally right?

File: torch/csrc/autograd/python_variable.cpp [hunk 5]
```diff
@@ -1692,7 +1725,11 @@ static void clear_slots(PyTypeObject* type, PyObject* self) {
   PyMemberDef* mp;
 
   n = Py_SIZE(type);
+#if PY_VERSION_HEX >= 0x030B0000
+  mp = type->tp_members;
```
Comment (albanD) [inline torch/csrc/autograd/python_variable.cpp:1729]: This can be used for all python versions no?

**Review by albanD — COMMENTED** [review]
Hey,

Thanks a lot for working on fixing these. I think we could make this change a lot nicer by defining the compatibilty macros as suggested in https://docs.python.org/3.11/whatsnew/3.11.html I left comments inline to that effect.

File: torch/csrc/autograd/python_variable.cpp (hunk 5 above)
Comment (cbalint13) [reply torch/csrc/autograd/python_variable.cpp:1708]: Right, so condition removed in this case.

File: torch/csrc/autograd/python_variable.cpp (hunk 3 above)
Comment (cbalint13) [reply torch/csrc/autograd/python_variable.cpp]: Right, modified as per your request.
For the track, the solution was from [python_function.cpp:L206](https://github.com/pytorch/pytorch/blob/30fb2c4abaaaa966999eab11674f25b18460e609/torch/csrc/autograd/python_function.cpp#L206)

File: torch/csrc/Stream.cpp (hunk 2 above)
Comment (cbalint13) [reply torch/csrc/Stream.cpp]: Added.

File: torch/csrc/autograd/python_variable.cpp (hunk 4 above)
Comment (cbalint13) [reply torch/csrc/autograd/python_variable.cpp]: Yes, we can have it for all versions **unconditionally**

**Comment by cbalint13** [comment]
> Hey,

Thank you @albanD for the prompt review !

> Thanks a lot for working on fixing these. I think we could make this change a lot nicer by defining the compatibilty macros as suggested in https://docs.python.org/3.11/whatsnew/3.11.html

> I left comments inline to that effect.

* Except this request to introduce compat-macros I updated all your suggested changes.

If want to move this PR quickly and have compat-macros feel free to fork out from this point.

Thank you.
~Cristian.

File: torch/csrc/autograd/profiler_python.cpp (hunk 1 above)
Comment (albanD) [reply torch/csrc/autograd/profiler_python.cpp:42]: > I would decline your idea to maintain compat declarations.

Ho interesting. Why do you think that having this functions defined once in a header and re-used in all the places we need is not as good as the specific ifdef that are currently in this PR?

Comment (cbalint13) [reply torch/csrc/autograd/profiler_python.cpp:42]: > > I would decline your idea to maintain compat declarations.
> 
> Ho interesting. Why do you think that having this functions defined once in a header and re-used in all the places we need is not as good as the specific ifdef that are currently in this PR?

Due to lack of time, I stop author here.
Feel free fork out for quick move over it.

Comment (albanD) [reply torch/csrc/autograd/profiler_python.cpp:42]: Ok, I was wondering if there were technical reasons.
I can take over this if you don't have time for the updates of course.

Thanks a lot for taking the time to make the first draft!

**Pull request closed without merging**